File: contents/cooley_tukey/code/julia/fft.jl
using FFTW

#simple DFT function
function DFT(x)
    N = length(x)

    # We want two vectors here for real space (n) and frequency space (k)
    n = 0:N-1
    k = n'
    transform_matrix = exp.(-2im*pi*n*k/N)
    return transform_matrix*x

end

# Implementing the Cooley-Tukey Algorithm
function cooley_tukey(x)
    N = length(x)

    if (N > 2)
        x_odd = cooley_tukey(x[1:2:N])
        x_even = cooley_tukey(x[2:2:N])
    else
        x_odd = x[1]
        x_even = x[2]
    end
    n = 0:N-1
    half = div(N,2)
    factor = exp.(-2im*pi*n/N)
    return vcat(x_odd .+ x_even .* factor[1:half],
                x_odd .- x_even .* factor[1:half])

end

function bitreverse(a::Array)
    # First, we need to find the necessary number of bits
    digits = convert(Int,ceil(log2(length(a))))

    indices = [i for i = 0:length(a)-1]

    bit_indices = []
    for i = 1:length(indices)
        push!(bit_indices, bitstring(indices[i]))
    end

    # Now stripping the unnecessary numbers
    for i = 1:length(bit_indices)
        bit_indices[i] = bit_indices[i][end-digits:end]
    end

    # Flipping the bits
    for i =1:length(bit_indices)
        bit_indices[i] = reverse(bit_indices[i])
    end

    # Replacing indices
    for i = 1:length(indices)
        indices[i] = 0
        for j = 1:digits
            indices[i] += 2^(j-1) * parse(Int, string(bit_indices[i][end-j]))
        end
       indices[i] += 1
    end

    b = [float(i) for i = 1:length(a)]
    for i = 1:length(indices)
        b[i] = a[indices[i]]
    end

    return b
end

function iterative_cooley_tukey(x)
    N = length(x)
    logN = convert(Int,ceil(log2(length(x))))
    bnum = div(N,2)
    stride = 0;

    x = bitreverse(x)

    z = [Complex(x[i]) for i = 1:length(x)]
    for i = 1:logN
       stride = div(N, bnum)
       for j = 0:bnum-1
           start_index = j*stride + 1
           y = butterfly(z[start_index:start_index + stride - 1])
           for k = 1:length(y)
               z[start_index+k-1] = y[k]
           end
       end
       bnum = div(bnum,2)
    end

    return z
end

function butterfly(x)
    N = length(x)
    half = div(N,2)
    n = [i for i = 0:N-1]
    half = div(N,2)
    factor = exp.(-2im*pi*n/N)

    y = [0 + 0.0im for i = 1:length(x)]

    for i = 1:half
        y[i] = x[i] + x[half+i]*factor[i]
        y[half+i] = x[i] - x[half+i]*factor[i]
    end

    return y
end

function main()
    x = rand(128)
    y = cooley_tukey(x)
    z = iterative_cooley_tukey(x)
    w = fft(x)
    if(isapprox(y, w))
        println("Recursive Cooley Tukey matches fft() from FFTW package.")
    end
    if(isapprox(z, w))
        println("Iterative Cooley Tukey matches fft() from FFTW package.")
    end
end

main()

File: contents/cooley_tukey/code/scala/fft.scala

case class Complex(val real: Double, val imag: Double) {
  def +(other: Complex): Complex = new Complex(this.real + other.real, this.imag + other.imag)
  def -(other: Complex): Complex = new Complex(this.real - other.real, this.imag - other.imag)
  
  def *(other: Complex): Complex = new Complex(
    this.real * other.real - this.imag * other.imag,
    this.real * other.imag + this.imag * other.real)
  
  def *(other: Double): Complex = new Complex(this.real * other, this.imag * other)
}

object Complex {
  def fromPolar(mag: Double, phase: Double): Complex = 
    new Complex(mag * Math.cos(phase), mag * Math.sin(phase))
}

object FT {

  /** Calculates a single DFT coefficient. */
  private def coefficient(n: Int, k: Int, ftLength: Int): Complex = 
    Complex.fromPolar(1.0, -2.0 * Math.PI * k * n / ftLength)
    
  /** Calculates one value of the DFT of signal */
  private def dftValue(signal: IndexedSeq[Double], k: Int): Complex = {
    // Multiply the signal with the coefficients vector
    val terms = for (i <- signal.indices) yield coefficient(i, k, signal.length) * signal(i)

    terms reduce { _ + _ }
  }
    
  def dft(signal: IndexedSeq[Double]): IndexedSeq[Complex] = 
    signal.indices map { dftValue(signal, _) }

  /** Combines the transforms of the even and odd indices */
  private def mergeTransforms(evens: IndexedSeq[Complex], odds: IndexedSeq[Complex]): IndexedSeq[Complex] = {
    val oddTerms = for (i <- odds.indices) 
      yield coefficient(1, i, 2 * odds.length) * odds(i)
    
    val pairs = evens.zip(oddTerms)

    (pairs map { case (e, o) => e + o }) ++ (pairs map { case (e, o) => e - o })
  }

  def cooleyTukey(signal: IndexedSeq[Double]): IndexedSeq[Complex] = signal.length match {
    case 2 => mergeTransforms(
      Vector(new Complex(signal(0), 0)),
      Vector(new Complex(signal(1), 0)))
    case _ => {
      // Split signal into even and odd indices and call cooleyTukey recursively on each.
      val evens = cooleyTukey(for (i <- 0 until signal.length by 2) yield signal(i))
      val odds = cooleyTukey(for (i <- 1 until signal.length by 2) yield signal(i))

      mergeTransforms(evens, odds)
    }
  }

  /** Reverses the bits in value. */
  private def reverseBits(value: Int, length: Int): Int = length match {
    case 1 => value
    case _ => {
      // Split bits in the middle.
      val lowerHalf = length / 2

      // The upper half will be longer if the number of bits is odd.
      val upperHalf = length - lowerHalf
      val mask = (1 << lowerHalf) - 1

      // Reverse each half recursively and then swap them.
      (reverseBits(value & mask, lowerHalf) << upperHalf) + 
        reverseBits(value >> lowerHalf, upperHalf)
    }
  }


  private def log2(x: Double): Double = Math.log(x) / Math.log(2.0)

  def bitReverseIndices(signal: IndexedSeq[Double]): IndexedSeq[Double] = {
    // Find the maximum number of bits needed.
    val bitLength = log2(signal.length).ceil.toInt

    for (i <- signal.indices)
      yield signal(reverseBits(i, bitLength))
  }

  private def butterfly(x1: Complex, x2: Complex, coeff: Complex): (Complex, Complex) = 
    (x1 + coeff * x2, x1 - coeff * x2)

  @scala.annotation.tailrec
  private def iterativeCooleyTukeyLoop(signal: IndexedSeq[Complex], dist: Int): IndexedSeq[Complex] = {
    // Distance between subsequent groups of butterflies
    val stride = 2 * dist
    val result = new Array[Complex](signal.length)

    for (groupStart <- 0 until signal.length by stride; i <- 0 until dist) {
      val index = groupStart + i
      val (r1, r2) = butterfly(
        signal(index),
        signal(index + dist),
        coefficient(1, i, stride))
      
      result(index) = r1
      result(index + dist) = r2
    }
    
    if (stride >= signal.length)
      result.toVector
    else
      iterativeCooleyTukeyLoop(result.toVector, dist * 2)
  }

  def iterativeCooleyTukey(signal: IndexedSeq[Double]): IndexedSeq[Complex] =
    iterativeCooleyTukeyLoop(bitReverseIndices(signal) map { new Complex(_, 0.0) }, 1)
}

object Main {

  private def approxEqual(a: IndexedSeq[Complex], b: IndexedSeq[Complex]): Boolean = {
    val diffs = a.zip(b) map { case (x, y) => Math.abs(x.real - y.real + x.imag - y.imag) }
    diffs map { _ < 1e-12 } reduce { _ && _ }
  }

  def main(args: Array[String]): Unit = {
    val signal = for (i <- 0 until 16) yield Math.random() * 2 - 1
    val x = FT.dft(signal)
    val y = FT.cooleyTukey(signal)
    val z = FT.iterativeCooleyTukey(signal)

    println("DFT and Cooley-Tukey approx. equal: " + approxEqual(x, y))
    println("DFT and iterative Cooley-Tukey approx. equal: " + approxEqual(x, z))
    println("Cooley-Tukey and iterative Cooley-Tukey approx. equal: " + approxEqual(y, z))
  }
}
